### PR TITLE
Initial work to add Web API Connect Endpoint

### PR DESCRIPTION
**Pull request by freman**
See: https://developer.spotify.com/web-api/web-api-connect-endpoint-reference/

Still needs some tests, more than willing to take feedback on naming or whatever.

### CODE REVIEW

**Comment by freman** [comment]
Not sure how to fix AppVeyer, it works in my checkout when I run it in the freman namespace (And when I rename the path to zmb3 namespace)

File: player.go [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+	"net/url"
+	"strconv"
+)
+
+// PlayerDevice contains information about a device that a user can play music on
+type PlayerDevice struct {
+	// ID of the device. This may be empty.
+	ID string `json:"id"`
+	// IsActive If this device is the currently active device.
+	IsActive bool `json:"is_active"`
```
Comment (zmb3) [inline player.go]: I would probably just call these `Active` and `Restricted`.

File: player.go [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+	// IsRestricted Whether controlling this device is restricted. At present if
+	// this is "true" then no Web API commands will be accepted by this device.
+	IsRestricted bool `json:"is_restricted"`
+	// Name The name of the device.
+	Name string `json:"name"`
+	// Type Device type, such as "Computer", "Smartphone" or "Speaker".
```
Comment (zmb3) [inline player.go]: Try to make the comments read more like sentences:

> Type of the device, such as "Computer", "Smartphone",  or "Speaker"

File: player.go [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+	Type string `json:"type"`
+	// Volume The current volume in percent.
+	Volume int `json:"volume_percent"`
+}
+
+// PlayerDeviceList is a list of devices
+type PlayerDeviceList struct {
```
Comment (zmb3) [inline player.go]: If you're not attaching any methods to `PlayerDeviceList`, I would prefer just to use `[]PlayerDevice` everywhere.  This way it's immediately obvious that we're returning a list (slice) of devices.

File: player.go [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+	// Device The device that is currently active
+	Device PlayerDevice `json:"device"`
+	// ShuffleState Shuffle is on or off
+	ShuffleState bool `json:"shuffle_state"`
+	// RepeatState off, track, context
+	RepeatState string `json:"repeat_state"`
+}
+
+// Context is the playback context
+type Context struct {
```
Comment (zmb3) [inline player.go]: Maybe we should call this `PlaybackContext` so that it doesn't seem related to `context.Context`

File: player.go [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+	// Progress into the currently playing track.
+	Progress int `json:"progress_ms"`
+	// IsPlaying If something is currently playing.
+	IsPlaying bool `json:"is_playing"`
+	// The currently playing track. Can be null.
+	Item *FullTrack `json:"Item"`
+}
+
+type Offset struct {
+	//Position is zero based and can’t be negative.
```
Comment (zmb3) [inline player.go]: Space after `//`

What are the units on position?

File: player.go [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+	Context Context `json:"context"`
+	// Progress into the currently playing track.
+	Progress int `json:"progress_ms"`
+	// IsPlaying If something is currently playing.
+	IsPlaying bool `json:"is_playing"`
+	// The currently playing track. Can be null.
+	Item *FullTrack `json:"Item"`
+}
+
+type Offset struct {
```
Comment (zmb3) [inline player.go]: `PlaybackOffset` ?

File: player.go [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+	//Position is zero based and can’t be negative.
+	Position int `json:"position,omitempty"`
+	// URI is a string representing the uri of the item to start at.
+	URI URI `json:"uri,omitempty"`
+}
+
+type PlayOptions struct {
+	// DeviceID The id of the device this command is targeting. If not
+	// supplied, the user's currently active device is the target.
+	DeviceID *string `json:"-"`
```
Comment (zmb3) [inline player.go]: Why `json:"-"`?
Is this in the form of a Spotify ID?  If so, we have a type for that.

File: player.go [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+// TransferPlayback transfers playback to a new device and determine if
+// it should start playing. This call requires authorization.
+//
+// Note that a value of false for the play parameter when also transferring
+// to another device_id will not pause playback. To ensure that playback is
+// paused on the new device you should send a pause command to the currently
+// active device before transferring to the new device_id.
+//
+// Requires the ScopeUserModifyPlaybackState in order to modify the player state
+func (c *Client) TransferPlayback(device_ids []string, play bool) error {
```
Comment (zmb3) [inline player.go]: Don't use underscores, call the argument `deviceIDs`.

Again, are these strings or Spotify IDs?

File: player.go [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+
+// Previous skips to the Previous track in the user's queue in the user's
+// currently active device. This call requires authorization.
+//
+// Requires the ScopeUserModifyPlaybackState in order to modify the player state
+func (c *Client) Previous() error {
+	return c.NextOpt(nil)
+}
+
+// PreviousOpt Like Previous but with more options
```
Comment (zmb3) [inline player.go]: PreviousOpt _is_ like Previous but with more options.

File: player.go [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+//
+// Requires the ScopeUserModifyPlaybackState in order to modify the player state
+func (c *Client) Repeat(state string) error {
+	return c.RepeatOpt(state, nil)
+}
+
+// RepeatOpt Like Repeat but with more options
+//
+// Only expects PlayOptions.DeviceID, all other options will be ignored
+func (c *Client) RepeatOpt(state string, opt *PlayOptions) error {
```
Comment (zmb3) [inline player.go]: What is `state`?
Should we remove `*PlayOptions` and accept an optional device ID instead?

File: player.go (hunk 1 above)
Comment (freman) [inline player.go]: Ok

File: player.go (hunk 2 above)
Comment (freman) [inline player.go]: Ok

File: player.go (hunk 3 above)
Comment (freman) [inline player.go]: Was more for the json.Unmarshal, but Ok

File: player.go (hunk 4 above)
Comment (freman) [inline player.go]: Ok

File: player.go (hunk 6 above)
Comment (freman) [inline player.go]: Ok

File: player.go (hunk 5 above)
Comment (freman) [inline player.go]: Surprised the linter doesn't pick that up.

RE Units: The api doesn't specify, in the context of their example it's the offset of a track in a playlist or list of uris, but I suspect it's also milliseconds when given a track.

File: player.go (hunk 7 above)
Comment (freman) [inline player.go]: Stops json.Marshal from marshalling the DeviceID which all requests expect in a get param not as part of the post/put - Being used for the convenience of it being optional in all the functions (so far only really one) that actually take these options vs using a seperate positional argument.

File: player.go (hunk 8 above)
Comment (freman) [inline player.go]: Yeh sorry, was tired

File: player.go (hunk 9 above)
Comment (freman) [inline player.go]: fixed

File: player.go (hunk 10 above)
Comment (freman) [inline player.go]: I was going for consistency (Ie: things suffixed 'opt' take opts...) but if you like I can do that.
